src/analyses/mallocWrapperAnalysis.ml

```diff
@@ -0,0 +1,80 @@
+(** An analysis that handles the case when malloc is called from a wrapper function all over the code. *)
+
+open Prelude.Ana
+open Analyses
+open GobConfig
+
+module Spec : Analyses.Spec =
+struct
+  include Analyses.DefaultSpec
+
+  module PL = Lattice.Flat (Basetype.ProgLines) (struct 
+    let top_name = "Unknown line"
+    let bot_name = "Unreachable line" 
+  end)
+
+  let name () = "mallocWrapper"
+  module D = PL
+  module G = Lattice.Unit
+  module C = D
+
+  module Q = Queries
+
+  (* transfer functions *)
+  let assign ctx (lval:lval) (rval:exp) : D.t =
+    ctx.local
+
+  let branch ctx (exp:exp) (tv:bool) : D.t =
+    ctx.local
+
+  let body ctx (f:fundec) : D.t =
+    ctx.local
+
+  let return ctx (exp:exp option) (f:fundec) : D.t =
+    ctx.local
+
+  let enter ctx (lval: lval option) (f:varinfo) (args:exp list) : (D.t * D.t) list =
+    let interestingfunctions = get_string_list "exp.malloc.wrappers" in
+    let calleofinterest = List.mem f.vname interestingfunctions in
+    let callectx = if calleofinterest then
+       if ctx.local = `Top then
+        `Lifted (MyCFG.getLoc ctx.node) 
+        else ctx.local
+      else D.top () in     
```

Comment:
The logic here is that a boring callee changes this to top, but wouldn't it be better to keep it at it's current value?
For example, if `ldv_malloc` that is listed as malloc wrapper calls some random wrapper `wrapper` that isn't listed as a malloc wrapper which then `malloc`s something. Right now, `wrapper` would forget the `ldv_malloc`s location and thus the inner `malloc` wouldn't have that precisely. Seems to me like we should remember the outermost malloc wrapper even if it goes through some things which we haven't listed as such.
Or does that cause some problem?

Reply:
I think the reasoning was avoiding doing a callstring-sensitive analysis of other functions in case the malloc wrapper also calls some other (non malloc-wrapper) functions. In case we don't go to top there, we will analyze all things transitively called from a wrapper function for each caller of the wrapper. This could potentially become very expensive with not much gain.

Reply:
My intuition would be that a malloc wrapper doesn't really call a whole bunch of different things that need lots of reanalysis. Right now, if multiple wrappers are used, we must list them all or we lose all the benefit. If the outermost wrapper was remembered through non-listed calls, then it's necessary to only have the outermost function listed. It would be prone against situations where another wrapper has been put into its implementation in some benchmarks.

We can leave it as is, but a comment in the code might be good to draw attention to the fact that it's supposed to currently forget wrappers in that case. If we have a somewhat sensible list of malloc wrappers for SV-COMP, I might want to do an experimental run the other way for comparison (of analysis times and precision).

Reply:
For now I wrote comments about this.